File: src/main/java/de/maggiwuerze/xdccloader/controller/ServerController.java
package de.maggiwuerze.xdccloader.controller;

import de.maggiwuerze.xdccloader.model.entity.Bot;
import de.maggiwuerze.xdccloader.model.entity.Server;
import de.maggiwuerze.xdccloader.model.forms.ServerForm;
import de.maggiwuerze.xdccloader.model.transport.ChannelTO;
import de.maggiwuerze.xdccloader.model.transport.ServerTO;
import de.maggiwuerze.xdccloader.service.ServerService;
import java.security.Principal;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.validation.ConstraintViolation;
import javax.validation.ConstraintViolationException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@Slf4j
@RequiredArgsConstructor
@RestController
class ServerController {

	private final ServerService serverService;

	/**
	 * @return a list of all servers
	 */
	@GetMapping("/servers/")
	public ResponseEntity<List<ServerTO>> getAllServers() {
		List<ServerTO> servers = serverService.list().stream().map(ServerTO::new).collect(Collectors.toList());;
		return new ResponseEntity(servers, HttpStatus.OK);
	}

	@PostMapping("/servers/")
	public ResponseEntity<?> addServer(@RequestBody ServerForm serverForm) {
		try {
			Server server = serverService.save(new Server(serverForm.getName(), serverForm.getServerUrl()));
			return new ResponseEntity("Download added successfully. id=[" + server.getId() + "]", HttpStatus.OK);
		} catch (ConstraintViolationException e) {
			log.error(e.getMessage());
			String errormessage = e.getConstraintViolations().stream().map(ConstraintViolation::getMessage).collect(Collectors.joining("\n"));
			return new ResponseEntity(errormessage, HttpStatus.BAD_REQUEST);
		}
	}

	/**
	 * @return an example server object to populate the attributes for server creation popover
	 */
	@GetMapping("/servers/example")
	public ResponseEntity<List<Server>> getExampleServer(Principal principal) {
		return new ResponseEntity(List.of(new Server()), HttpStatus.OK);
	}

	@DeleteMapping("/servers/")
	public ResponseEntity<?> delete( Long serverId) {
		try {
			serverService.delete(serverId);
			return new ResponseEntity("Server deleted successfully.", HttpStatus.OK);
		}catch (Exception e) {
			return new ResponseEntity("Server could not be deleted", HttpStatus.CONFLICT);
		}
	}
}

File: src/main/kotlin/de/maggiwuerze/xdccwebloader/controller/ServerController.kt
package de.maggiwuerze.xdccwebloader.controller

import de.maggiwuerze.xdccwebloader.persistence.entity.Server
import de.maggiwuerze.xdccwebloader.persistence.entity.ServerTO
import de.maggiwuerze.xdccwebloader.model.forms.ServerFormTO
import de.maggiwuerze.xdccwebloader.service.ServerService
import jakarta.validation.ConstraintViolationException
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.*

@RestController
@RequestMapping("server/")
internal class ServerController(private val serverService: ServerService) {

    val log = LoggerFactory.getLogger(this.javaClass.name)

    @GetMapping
    fun listServers(): ResponseEntity<List<ServerTO>> {
        return ResponseEntity(serverService.list().map { it.toTO() }.toList(), HttpStatus.OK)
    }

    @GetMapping("{id}")
    fun getServer(@PathVariable id: UUID): ResponseEntity<ServerTO> {
        serverService.findById(id)?.let {
            return ResponseEntity(it.toTO(), HttpStatus.OK)
        }
        return ResponseEntity(HttpStatus.NOT_FOUND)
    }

    @PostMapping
    fun createServer(@RequestBody serverFormTO: ServerFormTO): ResponseEntity<ServerTO> {
        try {
            serverService.save(Server(name = serverFormTO.name, serverUrl = serverFormTO.serverUrl)).let { server ->
                return ResponseEntity(server.toTO(), HttpStatus.OK)
            }
        } catch (e: ConstraintViolationException) {
            log.error(e.message)
            return ResponseEntity(HttpStatus.BAD_REQUEST)
        }
    }

    @PutMapping("{id}")
    fun updateServer(@PathVariable id: UUID, @RequestBody serverFormTO: ServerFormTO): ResponseEntity<ServerTO> =
        serverService.update(id, serverFormTO).toTO().let { ResponseEntity(it, HttpStatus.OK) }

    @DeleteMapping("{id}")
    fun deleteServer(@PathVariable id: UUID): ResponseEntity<*> {
        try {
            serverService.delete(id)
            return ResponseEntity("Server deleted successfully.", HttpStatus.OK)
        } catch (e: Exception) {
            return ResponseEntity("Server could not be deleted", HttpStatus.CONFLICT)
        }
    }
}